### PR TITLE
ImpliesAtom

### PR DESCRIPTION
**Pull request by jamestouri**
When working on the constraint validation at https://github.com/medplum/medplum/issues/2064, there are several key features in FHIRPath logic that hasn't been implemented. One of them includes `Implies` which is needed for several rule constraints

http://hl7.org/fhirpath/N1/#implies

### CODE REVIEW

File: packages/core/src/fhirpath/atoms.test.ts [hunk 1]
```diff
@@ -52,6 +52,20 @@ describe('Atoms', () => {
     expect(evalFhirPath('x | y', [])).toEqual([]);
   });
 
+  test.each([
+    ['true implies true', true],
+    ['true implies false', false],
+    ['true implies null', undefined],
```
Comment (codyebberson) [inline packages/core/src/fhirpath/atoms.test.ts:58]: Can you add test cases for the "empty" column on this page: https://build.fhir.org/ig/HL7/FHIRPath/

You might need to refactor the `test.each()` array to be able to handle the empty array case.

```ts
['true implies {}', []],
['false implies {}', [true]],
['{} implies false', []],
['{} implies {}', []],
```

**Review by codyebberson — COMMENTED** [review]
Nice, clean.  Thanks @jamestouri 

See note on some additional test cases from https://build.fhir.org/ig/HL7/FHIRPath/

**Comment by codyebberson** [comment]
(duplication warning is ok with me -- it would be a gross refactor for minimal gain)

File: packages/core/src/fhirpath/atoms.ts [hunk 2]
```diff
@@ -352,6 +352,35 @@ export class XorAtom extends InfixOperatorAtom {
...
+    super('implies', left, right);
+  }
+
+  eval(context: AtomContext, input: TypedValue[]): TypedValue[] {
+    const leftResult = this.left.eval(context, input);
+    const rightResult = this.right.eval(context, input);
+    if (leftResult.length === 0 && rightResult.length === 0) {
+      return [];
+    }
+    const leftValue = leftResult.length === 0 ? null : leftResult[0].value;
```
Comment (mattwiller) [inline packages/core/src/fhirpath/atoms.ts:372]: In the case where either argument collection has more than one element in it, this should probably produce an error to let the user know that the expression is incorrect.  In general, FHIRPath is pretty strict when coercing from a collection to a single value; that part of evaluation is covered explicitly here: http://hl7.org/fhirpath/N1/#singleton-evaluation-of-collections

File: packages/core/src/fhirpath/atoms.ts [hunk 3]
```diff
@@ -352,6 +352,35 @@ export class XorAtom extends InfixOperatorAtom {
...
+    const rightValue = rightResult.length === 0 ? null : rightResult[0].value;
+    if (rightValue === true || leftValue === false || (leftValue === null && rightValue === true)) {
+      return booleanToTypedValue(true);
+    }
+    if (leftValue === true && rightValue === false) {
+      return booleanToTypedValue(false);
+    }
+    return [];
```
Comment (mattwiller) [inline packages/core/src/fhirpath/atoms.ts:380]: You already have a case above that returns `[]`, which I think could be combined if you reorganize the conditions like so:
```ts
if (leftValue === false || rightValue === true) {
  return true;
} else if (leftValue === null || rightValue === null) {
  return [];
} else {
  return false;
}
```

File: packages/core/src/fhirpath/parse.ts [hunk 4]
```diff
@@ -210,7 +211,8 @@ export function initFhirPathParserBuilder(): ParserBuilder {
       (left, _, right) => new ArithemticOperatorAtom('mod', left, right, (x, y) => x % y)
     )
     .infixLeft('or', OperatorPrecedence.Is, (left, _, right) => new OrAtom(left, right))
-    .infixLeft('xor', OperatorPrecedence.Is, (left, _, right) => new XorAtom(left, right));
+    .infixLeft('xor', OperatorPrecedence.Is, (left, _, right) => new XorAtom(left, right))
+    .infixLeft('implies', OperatorPrecedence.Is, (left, _, right) => new ImpliesAtom(left, right));
```
Comment (mattwiller) [inline packages/core/src/fhirpath/parse.ts:215]: Why wouldn't this be `OperatorPrecedence.Implies`?

File: packages/core/src/fhirpath/atoms.ts [hunk 5]
```diff
@@ -352,6 +352,43 @@ export class XorAtom extends InfixOperatorAtom {
...
+    const leftResult = this.left.eval(context, input);
+    const rightResult = this.right.eval(context, input);
+    if (!this.isValidCollectionLength(leftResult, rightResult)) {
+      throw new Error(
+        `One or more collections are not of length 1 leftResult: ${leftResult.length} rightResult: ${rightResult.length}`
+      );
+    }
+    const leftValue = leftResult.length === 0 ? null : leftResult[0].value;
+    const rightValue = rightResult.length === 0 ? null : rightResult[0].value;
+    if (rightValue === true || leftValue === false || (leftValue === null && rightValue === true)) {
```
Comment (mattwiller) [inline packages/core/src/fhirpath/atoms.ts:376]: The rightmost part of this condition is superfluous; it can be rewritten equivalently as `rightValue === true || leftValue === false`

File: packages/core/src/fhirpath/atoms.ts [hunk 6]
```diff
@@ -352,6 +352,43 @@ export class XorAtom extends InfixOperatorAtom {
...
+      return [];
+    }
+    return booleanToTypedValue(false);
+  }
+
+  private isValidCollectionLength(leftResult: TypedValue[], rightResult: TypedValue[]): boolean {
+    if (leftResult.length > 1 || rightResult.length > 1) {
+      return false;
+    }
+    return true;
```
Comment (mattwiller) [inline packages/core/src/fhirpath/atoms.ts:388]: This can be rewritten more concisely as `return (leftResult.length <= 1 && rightValue.length <= 1);`

File: packages/core/src/fhirpath/atoms.ts (hunk 5 above)
Comment (jamestouri) [reply packages/core/src/fhirpath/atoms.ts:376]: Ah, that's right good catch

**Review by mattwiller — APPROVED** [review]
